Add a "Bookmarks" sidebar to Caja

thesquash opened the pull request:
One user requested to have a new sidebar added to Caja.  This sidebar shows only the user's own bookmarks.  When any bookmark in the sidebar is clicked on, Caja will open the bookmark in the current window and tab.  When the user middle-clicks the bookmark, Caja will open the bookmark in a new tab.

This feature was requested because the Places sidebar was a little too overcrowded for the requestor.

thesquash commented:
The original discussion, plus a screengrab of my patch in action, is at the Ubuntu MATE forum:

[](https://ubuntu-mate.community/t/feature-request-bookmarks-in-caja-sidepanel-pulldown/22260/10)

thesquash commented:
Thank you, @lukefromdc!  I did expect it to build correctly, because though I spent only 4 hours slapping it together, I did test it pretty thoroughly (I think).

I hope this passes the review.

lukefromdc commented:
Should I go ahead and merge this on one review, or does anyone else want to have a look at it?
Note the limitations mentioned in
https://ubuntu-mate.community/t/feature-request-bookmarks-in-caja-sidepanel-pulldown/22260/10

thesquash commented:
Thanks greatly for noticing, @rbuj.  I just made the change and committed again.

rbuj commented:
I don't know if it's a good idea to filter-out the built-in bookmarks, because if the user didn't add any bookmark it displays a blank box. It's only a personal assessment and other team members may consider as valid content.
```diff
diff --git a/src/caja-bookmarks-sidebar.c b/src/caja-bookmarks-sidebar.c
index b60c17b..0de3e26 100644
--- a/src/caja-bookmarks-sidebar.c
+++ b/src/caja-bookmarks-sidebar.c
@@ -94,30 +94,6 @@ G_DEFINE_TYPE_WITH_CODE (CajaBookmarksSidebarProvider, caja_bookmarks_sidebar_pr
                          G_IMPLEMENT_INTERFACE (CAJA_TYPE_SIDEBAR_PROVIDER,
                                  sidebar_provider_iface_init));
 
-
-static gboolean
-is_built_in_bookmark (CajaFile *file)
-{
-    gboolean built_in;
-    gint idx;
-
-    built_in = FALSE;
-
-    for (idx = 0; idx < G_USER_N_DIRECTORIES; idx++) {
-        /* PUBLIC_SHARE and TEMPLATES are not in our built-in list */
-        if (caja_file_is_user_special_directory (file, idx)) {
-            if (idx != G_USER_DIRECTORY_PUBLIC_SHARE &&  idx != G_USER_DIRECTORY_TEMPLATES) {
-                built_in = TRUE;
-            }
-
-            break;
-        }
-    }
-
-    return built_in;
-}
-
-
 static void
 update_bookmarks (CajaBookmarksSidebar *sidebar)
 {
@@ -158,13 +134,6 @@ update_bookmarks (CajaBookmarksSidebar *sidebar)
         root = caja_bookmark_get_location (bookmark);
         file = caja_file_get (root);
 
-        if (is_built_in_bookmark (file)) {
-            g_object_unref (root);
-            caja_file_unref (file);
-            continue;
-        }
-        
-
         surface = caja_bookmark_get_surface (bookmark, GTK_ICON_SIZE_MENU);
         name = caja_bookmark_get_name (bookmark);
         gtk_list_store_append (store, &iter);
```

On the other hand, there are many blank lines that are not needed as well as some trailing blank space.

thesquash commented:
Thank you, @rbuj, for your careful searching for my coding flaws.

I'll address the latter issue, the code formatting, first.  I'm so sorry you had to read that garbage code of mine, kind of ruthlessly copied from the ancient History sidebar code which wasn't properly indented.  I cleaned up the code a little and also found some trailing spaces in the header file for the sidebar.  (By the way, I must say the new feature in Pluma 1.24 which shows trailing spaces is indispensable.  I thank anyone and everyone who added that feature.)

The other issue you brought up, the blank sidebar, I agree with, retaining one reservation:  I still (personally) think that the "built-in" XDG bookmarks should not appear in the list, because the whole idea of this sidebar is to reduce clutter; if clutter were allowable, then we'd just use the Places sidebar.

However, I believe that it's stupid to leave the user scratching their head wondering why the sidebar is empty, so I added a few lines of code which detect if there are no user-defined bookmarks and adds a row that says "No bookmarks defined" in italics if so.  Again, I kind of stole the translations that were already present in Caja, but I think that's a benefit.

Additionally I've added code which prevents the user from selecting the "No bookmarks defined" row, so that users will not get confused about entering a bookmarked directory called "No bookmarks defined".

Thanks again everybody.

raveit65 commented:
This looks a bit weird with Traditional themes.
How does it look with Adwaita and HighContrast themes from Gtk ?
Here we can't fix it with themes settings.

rbuj commented:
The GtkTreeView only looks weird with *Menta and *Traditional themes.

Test: `sudo GTK_DEBUG=interactive caja`

<img width="1120" alt="Captura de Pantalla 2020-07-25 a les 17 32 32" src="https://user-images.githubusercontent.com/10171411/88460548-4883b580-ce9d-11ea-9c3d-ceaba08d0ec5.png">

raveit65 commented:
OK, i will take a look into it when merged.
Merge it when it works for you.
